controller/resource.go

```diff
@@ -49,47 +48,146 @@ type TaskLister struct {
 	// SecretPrefix is the prefix to determine names of resources in Consul or AWS.
 	SecretPrefix string
 
-	// Log is the logger for the TaskLister.
+	// Log is the logger for the ServiceStateLister.
 	Log hclog.Logger
 }
 
-// List lists all tasks for the Cluster.
-func (t TaskLister) List() ([]Resource, error) {
+// List returns a mapping from inferred service names to the the ACL tokens,
+// ECS tasks and existance of a Consul service.
+func (s ServiceStateLister) List() ([]Resource, error) {
 	var resources []Resource
+	buildingResources := make(map[string]*ServiceInfo)
+
+	tasks, err := s.fetchECSTasks()
+
+	if err != nil {
+		return resources, err
+	}
+
+	for name, tasks := range tasks {
+		if _, ok := buildingResources[name]; !ok {
+			buildingResources[name] = &ServiceInfo{
+				SecretsManagerClient: s.SecretsManagerClient,
+				ConsulClient:         s.ConsulClient,
+				Cluster:              s.Cluster,
+				Log:                  s.Log,
+				SecretPrefix:         s.SecretPrefix,
+				ServiceName:          name,
+				ServiceState:         ServiceState{ConsulECSTasks: tasks},
+			}
+		} else {
+			buildingResources[name].ServiceState.ConsulECSTasks = tasks
+		}
+	}
+
+	services, _, err := s.ConsulClient.Catalog().Services(nil)
+
+	if err != nil {
+		return resources, err
+	}
+
+	for name := range services {
+		if name == "consul" {
+			continue
+		}
+
+		if _, ok := buildingResources[name]; !ok {
+			buildingResources[name] = &ServiceInfo{
```

Comment:
> Excluding services to be running on this specific ECS cluster could lead to a case where tokens are deleted for a service that still exists elsewhere.

Hmm... I think we should limit this controller to managing tokens only for tasks in this cluster.

* If someone runs two ECS clusters, then (with current architecture) both clusters have an ACL controller. If there are tasks for a service in both clusters, then we should see two tokens created for the service, one per cluster/controller.
* If there are service instances running elsewhere (outside of ECS), those service instances should create a separate service token rather than reusing a controller-managed token.

From that perspective, controller-managed tokens should not be in use elsewhere, so it should be safe for a controller to delete a token that it manages (once there are no tasks in the local ECS cluster using the token).

Reply:
I updated this to only take tasks and ACL tokens into account. I'm super interested in what you all think